run-sandbox.1
.\" Automatically generated by Pandoc 3.1.8
.\"
.TH "sandbox-manager" "1" "" "Version 1.0.0" "sandbox-manager Usage Guide"
.SH NAME
\f[B]sandbox-manager\f[R] - A python script to manage bubblewrap (bwrap)
sandboxes using YAML configuration files.
.SH SYNOPSIS
run-sandbox [\f[B]-h\f[R]] [\f[B]--flatten\f[R]] [\f[B]--blocking\f[R]]
[\f[B]--run\f[R] \f[I]EXECUTABLE\f[R]] [\f[B]--search-in\f[R]
\f[I]DIR\f[R]] \f[I]filename\f[R]
.SH DESCRIPTION
To use the \f[CR]sandbox\f[R] command, provide a valid configuration
file as the \f[CR]filename\f[R] argument.
sandbox-manager will search for configuration files in the following
places, in the following order:
.IP "1." 3
In the directories specified in each instance of the
\f[I]\[en]search-in\f[R] command-line option.
.IP "2." 3
In the directories specified in the \f[B]SANDBOX_CONFIG_DIRS\f[R]
environment variable, with each directory seperated by colons.
.IP "3." 3
In the \f[B]default_configs\f[R] directory next to
\f[B]run-sandbox.py\f[R].
.SH OPTIONS
\f[B]filename\f[R]
.PD 0
.P
.PD
The name of the configuration file to run, without the file extension.
.PP
\f[B]\[en]help -h\f[R]
.PD 0
.P
.PD
Show a help message and exit.
.PP
\f[B]\[en]flatten -f\f[R]
.PD 0
.P
.PD
Parse the given configuration file and its dependencies, print out a new
configuration file with the dependencies integrated, then exit.
Useful for determining exactly what a program will be given access to.
.PP
\f[B]\[en]blocking -b\f[R]
.PD 0
.P
.PD
If set, the script will wait for the sandbox process to terminate before
terminating itself.
Useful when running a shell so that it runs in the foreground.
.PP
\f[B]\[en]run\f[R] \f[I]EXECUTABLE\f[R]
.PD 0
.P
.PD
A program to run instead of the one specified in the given config file.
Useful for running a shell in the sandboxed application\[cq]s
environment.
.PP
\f[B]\[en]search-in\f[R] \f[I]DIR\f[R] \f[B]-s\f[R] \f[I]DIR\f[R]
.PD 0
.P
.PD
A directory to search for config files in.
Can be specified multiple times.
.SH CONFIGURATION
Configuration files are defined using the YAML format.
They require exactly one \f[B]run\f[R] key (if used as the
\f[I]filename\f[R] argument), but a \f[B]name\f[R] key is also strongly
recommended.
Any values of a given configuration option will be passed to a shell, so
it is important that only trusted configuration files are used (and
inherited).
.PP
The valid configuration options are:
.SS \f[B]name\f[R]: \f[I]string\f[R]
Defines the name of the application.
Copied (with whitespace removed) to the \f[CR]appName\f[R] environment
variable when running bubblewrap.
For a version with whitespace included, use the \f[CR]appNameWspace\f[R]
variable.
If \f[CR]name\f[R] is not specified, the sandbox will run, but a warning
will be printed to the console.
.SS \f[B]run\f[R]: \f[I]executable\f[R]
Defines the application to run.
Since it is passed to a shell, it can be called in any way normally
possible from a shell (i.e.\ as a path or executable name).
.SS \f[B]inherit\f[R]: \f[I]list\f[R]
Takes a list of sandbox names to inherit.
Inherited sandboxes have their \f[B]preprocess\f[R] and
\f[B]permission\f[R] categories merged with the parent sandbox\[cq]s
respective categories.
Inheritance is performed recursively, so inherited sandboxes can also
have their own inheritances, but it is possible to get into infinite
loops, so use inheritance carefully.
The \f[B]run\f[R] configuration option will not be inherited, and a
warning will be printed to the console if an inherited sandbox contains
it.
.SS \f[B]preprocess\f[R]
Used to perform some operation before running the sandbox.
Currently, only the \f[B]mkdir\f[R] option exists.
.PP
\f[B]mkdir\f[R]: \f[I]directory\f[R]
.PD 0
.P
.PD
Create a new directory at \f[I]directory\f[R].
Also creates parent directories if they don\[cq]t already exist.
.SS \f[B]permissions\f[R]
Defines permissions for the sandbox.
Unless otherwise specified, each option under these categories contains
a list of the specified type (e.g.\ a list of paths).
It contains the following options:
.PP
\f[B]filesystem\f[R]
.PD 0
.P
.PD
Defines filesystem permissions.
Can contain the following options:
.RS
.PP
\f[B]bind\f[R]: \f[I]path\f[R]
.PD 0
.P
.PD
Allows (binds) a file or directory in the sandbox, with both read and
write access for applications within it.
.RE
.RS
.PP
\f[B]bind-to\f[R]: \f[I]path1\f[R] \f[I]path2\f[R]
.PD 0
.P
.PD
Allows (binds) a file or directory \f[I]path1\f[R] in the sandbox with
read and write access, but places it at the location \f[I]path2\f[R]
within the sandbox.
.RE
.RS
.PP
\f[B]ro-bind\f[R]: \f[I]path\f[R]
.PD 0
.P
.PD
Allows (binds) a file or directory in the sandbox as read-only.
.RE
.RS
.PP
\f[B]ro-bind-to\f[R]: \f[I]path1\f[R] \f[I]path2\f[R]
.PD 0
.P
.PD
Allows (binds) a file or directory \f[I]path1\f[R] in the sandbox as
read-only, but places it at the location \f[I]path2\f[R] within the
sandbox.
.RE
.RS
.PP
\f[B]bind-devices\f[R]: \f[I]path\f[R]
.PD 0
.P
.PD
Allows (binds) a device file, or a directory containing device files,
into the sandbox.
Can be used on non-device files, but this option is required for device
files to work properly in the sandbox.
.RE
.RS
.PP
\f[B]bind-devices-to\f[R]: \f[I]path1\f[R] \f[I]path2\f[R]
.PD 0
.P
.PD
Allows (binds) a device file or directory containing device files
\f[I]path1\f[R], but places it at the location \f[I]path2\f[R] within
the sandbox.
.RE
.RS
.PP
\f[B]link\f[R]: \f[I]path1\f[R] \f[I]path2\f[R]
.PD 0
.P
.PD
Creates a symlink in the sandbox, redirecting read/writes at
\f[I]path2\f[R] to \f[I]path1\f[R].
.RE
.RS
.PP
\f[B]new-dev\f[R]: \f[I]path\f[R]
.PD 0
.P
.PD
Creates a new device filesystem at \f[I]path\f[R] containing
\f[I]/dev/null\f[R], \f[I]/dev/zero\f[R], \f[I]/dev/urandom\f[R] and
other commonly needed device files.
.RE
.RS
.PP
\f[B]new-tmpfs\f[R]: \f[I]path\f[R]
.PD 0
.P
.PD
Creates a new tmpfs (temporary filesystem) at \f[I]path\f[R].
.RE
.RS
.PP
\f[B]new-proc\f[R]: \f[I]path\f[R]
.PD 0
.P
.PD
Creates a procfs containing all running processes.
If \f[B]share-pid\f[R] is not set, this will only contain processes
running within the sandbox.
.RE
.RS
.PP
\f[B]create-files\f[R]: \f[I]path\f[R]: \f[I]data\f[R]
.PD 0
.P
.PD
Creates a file in the sandbox at \f[I]path\f[R], containing the string
\f[I]data\f[R].
This option should contain key-value pairs of the form
\[lq]\f[I]path\f[R]: \f[I]data\f[R]\[rq].
.RE
.PP
Every option containing \f[B]bind\f[R] can be appended with the
`\f[B]-opt\f[R]' suffix, indicating that bubblewrap should silently fail
if the file or directory doesn\[cq]t exist.
.PP
\f[B]namespaces\f[R]
.PD 0
.P
.PD
Defines namespace permissions (e.g.\ user namespaces).
This option is a list of namespaces to share.
.RS
.PP
\f[B]share-user\f[R]
.PD 0
.P
.PD
Shares the user namespace.
Enabling this allows applications to use the user configurations of the
host system (i.e.\ a new user can\[cq]t be created with an already
existing uid).
.RE
.RS
.PP
\f[B]share-ipc\f[R]
.PD 0
.P
.PD
Shares the ipc namespace.
Enabling this allows sandboxed applications to communicate with other
(unsandboxed) applications on the system.
.RE
.RS
.PP
\f[B]share-pid\f[R]
.PD 0
.P
.PD
Shares the pid namespace.
Enabling this allows sandboxed applications to view and interact with
other (unsandboxed) processes on the system.
Note that sandboxed applications can still see other applications within
the sandbox, even without this option.
.RE
.RS
.PP
\f[B]share-network\f[R]
.PD 0
.P
.PD
Shares the network namespace.
Enabling this allows applications to access the network.
.RE
.RS
.PP
\f[B]share-hostname\f[R]
.PD 0
.P
.PD
Shares the uts namespace.
Enabling this allows applications to change the system\[cq]s hostname
for all processes (assuming they have permission to do so).
Note that even without this option, the system\[cq]s hostname is shared
with the sandboxed application.
.RE
.RS
.PP
\f[B]share-cgroup\f[R]
.PD 0
.P
.PD
Shares the cgroup namespace.
.RE
.PP
\f[B]environment\f[R]
.PD 0
.P
.PD
Defines environment variables passed to (or created in) the sandbox.
.RS
.PP
\f[B]copyenv\f[R] \f[I]env\f[R]
.PD 0
.P
.PD
Copies the environment variable \f[I]env\f[R] into the sandbox.
.RE
.RS
.PP
\f[B]setenv\f[R] \f[I]env\f[R] \f[I]value\f[R]
.PD 0
.P
.PD
Sets the environment variable \f[I]env\f[R] to \f[I]value\f[R] in the
sandbox.
.RE
.PP
\f[B]dbus\f[R]
.PD 0
.P
.PD
Defines access to D-Bus services.
Uses \f[I]xdg-dbus-proxy\f[R] to filter queries.
.RS
.PP
\f[B]see\f[R] \f[I]service\f[R]
.PD 0
.P
.PD
Allows an application to see D-Bus service \f[I]service\f[R] on the bus
(i.e.\ get its name and ID), but not communicate with it.
.RE
.RS
.PP
\f[B]talk\f[R] \f[I]service\f[R]
.PD 0
.P
.PD
Allows an application to send method calls and recieve signals from the
service.
.RE
.RS
.PP
\f[B]own\f[R] \f[I]service\f[R]
.PD 0
.P
.PD
Allows an application to own the name of the service.
.RE
.SH AUTHORS
catcraft (\f[I]https://github.com/CatCraftYT\f[R]).


src/run-sandbox.py
from classes.sandbox import Sandbox
from classes.config_loader import ConfigLoader
from yaml import safe_dump
from argparse import ArgumentParser
import sys
import os

argparser = ArgumentParser(
    prog="run-sandbox",
    description="Python script for managing bubblewrap (bwrap) sandboxes.",
    epilog="Configuration files will be read from directories contained within \
    the '--search-in' argument(s), or from the 'SANDBOX_CONFIG_DIRS' environment \
    variable (seperated by colons), in that order. \
    Read the man page for more information on configuration and security."
)
argparser.add_argument(
    "--flatten", "-f",
    action="store_true",
    default=False,
    help="Parse the given configuration file and its dependencies, \
    print out a new configuration file with the dependencies integrated, then exit. \
    Useful for determining exactly what a program will be given access to."
)
argparser.add_argument(
    "--run",
    metavar="EXECUTABLE",
    default=None,
    help="A program to run instead of the one specified in the given config file. \
    Useful for running a shell in your program's environment."
)
argparser.add_argument(
    "--search-in", "-s",
    action="append",
    default=[],
    metavar="DIR",
    help="A directory to search for config files in. Can be specified multiple times."
)
argparser.add_argument(
    "filename",
    help="The name of the configuration file to run, without the file extension."
)
argparser.add_argument(
    "args",
    nargs="*",
    help="Extra arguments to pass to the executable running in the sandbox. Use quotation \
    marks around arguments with hyphens."
)

# Path of this script in the filesystem
script_path = os.path.abspath(os.path.dirname(__file__))
args = argparser.parse_args()

search_paths = args.search_in + os.environ.get("SANDBOX_CONFIG_DIRS", "").split(":") + [os.path.join(script_path, "default_configs/")]
# Remove empty strings and lists
search_paths = [i for i in search_paths if i]

config_loader = ConfigLoader(search_paths)
config_loader.load(args.filename)

if args.flatten:
    safe_dump(config_loader.config, sys.stdout, sort_keys=False)
    sys.exit(0)

sandbox = None
sandbox = Sandbox(config_loader.config, extra_args=args.args)

if args.run:
    sandbox.executable = args.run

sandbox.run()
